File: sections/01_complexity_io_constraints/problems/c_token_budget/solution.cpp
#include <bits/stdc++.h>
using namespace std;

using int64 = long long;
using i128 = __int128_t;

int ceil_log2_for_estimate(int64 n) {
    int ans = 0;
    int64 value = 1;
    while (value < n) {
        value <<= 1;
        ++ans;
    }
    return max(1, ans);
}

bool fits(i128 cost, int64 budget) {
    return cost <= static_cast<i128>(budget);
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int Q;
    cin >> Q;
    while (Q--) {
        int64 n, budget;
        cin >> n >> budget;
        int64 log_estimate = ceil_log2_for_estimate(n);

        if (fits(static_cast<i128>(n) * n * n, budget)) {
            cout << "cubic\n";
        } else if (fits(static_cast<i128>(n) * n, budget)) {
            cout << "quadratic\n";
        } else if (fits(static_cast<i128>(n) * log_estimate, budget)) {
            cout << "nlogn\n";
        } else if (fits(n, budget)) {
            cout << "linear\n";
        } else {
            cout << "none\n";
        }
    }

    return 0;
}

File: sections/01_complexity_io_constraints/problems/c_token_budget/solution.py
"""Reference solution for C. Token Budget."""

import sys


def ceil_log2_for_estimate(n: int) -> int:
    return max(1, (n - 1).bit_length())


def classify(n: int, budget: int) -> str:
    log = ceil_log2_for_estimate(n)
    estimates = [
        ("cubic", n * n * n),
        ("quadratic", n * n),
        ("nlogn", n * log),
        ("linear", n),
    ]
    for label, cost in estimates:
        if cost <= budget:
            return label
    return "none"


def main() -> None:
    data = list(map(int, sys.stdin.buffer.read().split()))
    if not data:
        return
    q = data[0]
    out: list[str] = []
    idx = 1
    for _ in range(q):
        n = data[idx]
        budget = data[idx + 1]
        idx += 2
        out.append(classify(n, budget))
    sys.stdout.write("\n".join(out))


if __name__ == "__main__":
    main()

